### ISSUE TITLE
strict-type-predicates: false positive for mapped template types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#3667]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.7.0-dev.20180110
- __Running TSLint via__: (pick one) CLI & VS Code

#### TypeScript code being linted

```typescript
const testDeepReadonly = <TThing extends { optional?: string }>(thing: Readonly<TThing>): string => {
	if (thing.optional === undefined) { // [strict-type-predicates]: Expression is always false.
		return "undefined";
	}

	return thing.optional;
};
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "strict-type-predicates": true,
    }
}
```

#### Actual behavior

TSLint assumes `thing.optional` is a (concrete?) real type, rather than extending `?: string`.

#### Expected behavior

It shouldn't.

This might be related to #2944 and fixed by #3417. I haven't looked into it much.

**Comment by adidahiya** [comment]
feel free to comment if this is still an issue for you and i'll re-open.
